src/proxy.ts

```diff
@@ -0,0 +1,460 @@
+              {
+                headers: req.headers,
+                data: requestData,
+                url: req.url,
+                params: new URLSearchParams(req.url),
+              },
+              {
+                headers: proxyRes.headers,
+                data: responseData,
+                statusCode: proxyRes.statusCode || 200,
+              },
+            );
+            res.writeHead(modifiedResponse.statusCode || 200, modifiedResponse.headers);
+            res.end(JSON.stringify(modifiedResponse.data));
+          });
+        } catch (e: unknown) {
+          res.writeHead(500, proxyRes.headers);
+          res.end(`Proxy error: ${e}`);
+        }
+      });
+    });
+
+    req.on('data', (chunk) => requestChunks.push(chunk));
+  });
+
+  private serverIsRunning = false;
+  private port = 21345;
+
+  constructor(realtimeClientOptions: Ably.ClientOptions) {
+    this.realtimeClientOptions = realtimeClientOptions;
+
+    this.server.on('upgrade', (req, socket, head) => {
+      this.wss.handleUpgrade(req, socket, head, (ws) => {
+        const targetWs = new WebSocket(this.targetRealtimeUrl);
+        const deferredValue = CompletableDeferred<void>();
+        targetWs.on('open', () => deferredValue.complete());
+
+        ws.on('message', (message) => {
+          let data: ProtocolMessage;
+          try {
+            data = JSON.parse(message.toString());
+          } catch (e: unknown) {
+            console.error(e);
+            return;
+          }
+
+          this.realtimeQueue.enqueue(async () => {
+            await deferredValue.get();
+            const modifiedMessage = await this.applyRealtimeMiddlewares(data, 'outgoing');
+            if (modifiedMessage) {
+              targetWs.send(JSON.stringify(modifiedMessage));
+            }
+          });
+        });
+
+        targetWs.on('message', (message) => {
+          let data: ProtocolMessage;
+          try {
+            data = JSON.parse(message.toString());
+          } catch (e: unknown) {
+            console.error(e);
+            return;
+          }
+
+          if (data.action === 4) {
+            this.connectionIdToWs.set(data.connectionId!!, ws);
+          }
+
+          this.realtimeQueue.enqueue(async () => {
+            const modifiedMessage = await this.applyRealtimeMiddlewares(data, 'incoming');
+            if (modifiedMessage) {
+              ws.send(JSON.stringify(modifiedMessage));
+            }
+          });
+        });
+
+        ws.on('close', () => {
+          targetWs.close();
+          this.closeWs(ws);
+        });
+        targetWs.on('close', () => this.closeWs(ws));
+      });
+    });
+  }
+
+  async observeNextRequest<T = any>(filter?: (req: Request<T>) => boolean): Promise<Request<T>> {
+    const deferredValue = CompletableDeferred<Request<T>>();
+    const unregister = this.registerRestMiddleware(async (req, res) => {
+      if (!filter || filter(req)) deferredValue.complete(req);
+      return res;
```

Comment:
```suggestion
      return req;
```

shouldn't it return request?